core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java

```diff
@@ -2057,15 +2076,264 @@ public void ready_then_transient_failure_again() {
+
+
+    // Refail the first one, after third time should refreshNameResolution
+    stateListener.onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    inOrder.verify(mockHelper, never()).refreshNameResolution();
+    stateListener.onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    inOrder.verify(mockHelper, never()).refreshNameResolution();
+    stateListener.onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    inOrder.verify(mockHelper).refreshNameResolution();
+  }
+
+
+  @Test
+  public void happy_eyeballs_pick_pushes_index_over_end() {
+    Assume.assumeTrue(enableHappyEyeballs); // This test is only for happy eyeballs
+
+    InOrder inOrder = inOrder(mockHelper, mockSubchannel1, mockSubchannel2, mockSubchannel3);
+    Status error = Status.UNAUTHENTICATED.withDescription("simulated failure");
+
+    List<EquivalentAddressGroup> addrs =
+        Lists.newArrayList(servers.get(0), servers.get(1), servers.get(2));
+    Subchannel[] subchannels = new Subchannel[] {mockSubchannel1, mockSubchannel2, mockSubchannel3};
+    SubchannelStateListener[] listeners = new SubchannelStateListener[subchannels.length];
+    loadBalancer.acceptResolvedAddresses(
+        ResolvedAddresses.newBuilder().setAddresses(addrs).setAttributes(affinity).build());
+    forwardTimeByConnectionDelay(2);
+    for (int i = 0; i < subchannels.length; i++) {
+      inOrder.verify(subchannels[i]).start(stateListenerCaptor.capture());
+      listeners[i] = stateListenerCaptor.getValue();
+      listeners[i].onSubchannelState(ConnectivityStateInfo.forNonError(IDLE));
+    }
+    assertEquals(IDLE, loadBalancer.getConcludedConnectivityState());
+
+    loadBalancer.acceptResolvedAddresses(
+        ResolvedAddresses.newBuilder().setAddresses(addrs).setAttributes(affinity).build());
+    inOrder.verify(mockHelper).updateBalancingState(eq(IDLE), pickerCaptor.capture());
+    SubchannelPicker requestingPicker = pickerCaptor.getValue();
+
+    // First pick moves index to addr 2
+    PickResult pickResult = requestingPicker.pickSubchannel(mockArgs);
+    assertEquals("RequestConnectionPicker", requestingPicker.getClass().getSimpleName());
+    assertEquals(PickResult.withNoResult(), pickResult);
+    assertEquals(IDLE, loadBalancer.getConcludedConnectivityState());
+
+    // Second pick moves index to addr 3
+    pickResult = requestingPicker.pickSubchannel(mockArgs);
+    assertEquals(PickResult.withNoResult(), pickResult);
+
+    // Sending TF state to one subchannel pushes index past end, but shouldn't do anything
+    listeners[2].onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    inOrder.verify(mockHelper, never()).updateBalancingState(eq(TRANSIENT_FAILURE), any());
+
+    // Put the LB into TF
+    listeners[0].onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    listeners[1].onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
```

Comment:
One example is here, say we have 3 subchannels(TF, TF, TF). Index points to invalid.
Then we have address update with exactly the same addresses. So we reset FirstPass and Index = 0. 
We start do requestConnection() which iterates all 3 addresses. 
After that we should have a `refreshNameResolution` but current behaviour won't, because that logic in only in the listener, which is not sufficient. Right? I understand this may complicate the logic, so if we do not have a simple solution we should comment.

Reply:
Since the subchannels are in TF, each of them is retrying their connections independently of the load balancer.  We reset firstPass to true, so as soon as one of those retries gets a TF response we'll call refreshNameResolution.

Reply:
Talked offline about the (potentially delayed, but purposely) `refreshNameResolution` behaviour when all subchannels are in TF upon address update, or part of the subchannels are in TF upon address update. This looks fine to me.